backend/internal/domain/services/task_list_service.go

```diff
@@ -0,0 +1,195 @@
+package services
+
+import (
+	"context"
+	"net/http"
+	"shodo/internal/domain/helpers"
+	"shodo/internal/repository"
+	"shodo/models"
+)
+
+const (
+	kDefaultTaskListTitle = "Shoppings"
+	kNotAllowed           = "operation permitted only for list owner or shared users"
+)
+
+type TaskListService struct {
+	TaskListRepository    repository.TaskList
+	UsersRepository       repository.Users
+	AuthenticationService *AuthenticationService
+}
+
+func (this *TaskListService) CreateDefaultTaskList(ctx context.Context, username, ownerId string) {
+	this.TaskListRepository.CreateTaskList(ctx, &models.TaskList{Title: username + " " + kDefaultTaskListTitle, Owner: ownerId})
+}
+
+func (this *TaskListService) CreateTaskList(ctx context.Context, list *models.TaskList) {
+	this.TaskListRepository.CreateTaskList(ctx, list)
+}
+
+func (this *TaskListService) GetTaskLists(ctx context.Context, userToken string) ([]models.TaskListShort, error) {
+	userId, err := helpers.GetUserIdFromToken(userToken)
+	if err != nil {
+		return nil, err
+	}
+
+	return this.TaskListRepository.GetTaskLists(ctx, userId)
+}
+
+func (this *TaskListService) GetTaskList(ctx context.Context, listId *string, userToken string) (models.TaskList, *models.Error) {
+	isListExists, err := this.TaskListRepository.CheckTaskListExists(ctx, *listId)
+	if !isListExists {
+		return models.TaskList{}, &models.Error{Code: http.StatusNotFound, Message: "list not found"}
+	}
+
+	isEditListAllowed, err := this.IsEditListAllowed(ctx, listId, userToken)
+	if err != nil {
+		return models.TaskList{}, &models.Error{Code: http.StatusInternalServerError, Message: err.Error()}
+	}
+
+	if !isEditListAllowed {
+		return models.TaskList{}, &models.Error{Code: http.StatusForbidden, Message: kNotAllowed}
+	}
+
+	resp, err := this.TaskListRepository.GetTaskList(ctx, listId)
+	if err != nil {
+		return models.TaskList{}, &models.Error{Code: http.StatusInternalServerError, Message: err.Error()}
+	}
+
+	return resp, nil
+}
+
+func (this *TaskListService) AddTaskToList(ctx context.Context, listId *string, task *models.Task, userToken string) (*string, *models.Error) {
+	isListExists, err := this.TaskListRepository.CheckTaskListExists(ctx, *listId)
+	if !isListExists {
+		return nil, &models.Error{Code: http.StatusNotFound, Message: "list not found"}
+	}
+
+	isEditListAllowed, err := this.IsEditListAllowed(ctx, listId, userToken)
+	if err != nil {
+		return nil, &models.Error{Code: http.StatusInternalServerError, Message: err.Error()}
+	}
+
+	if !isEditListAllowed {
+		return nil, &models.Error{Code: http.StatusForbidden, Message: kNotAllowed}
+	}
+
+	taskId, err := this.TaskListRepository.AddTaskToList(ctx, listId, task)
+	if err != nil {
+		return nil, &models.Error{Code: http.StatusInternalServerError, Message: err.Error()}
+	}
+
+	return taskId, nil
+}
+
+func (this *TaskListService) RemoveTaskFromList(ctx context.Context, listId *string, taskId *string, userToken string) *models.Error {
```

Comment:
If you need to signal an error, you need to return a type `error`
It can lead to subtle problems later and you should never base your errors on strong types.

You can always inspect them with `errors.Is` or `errors.As`

Reply:
If you use errors.As, you can source error type into your structure easily. Something like
```go
var modelErr *models.Error
if errors.As(err, &modelErr) {
// your error is sourced into modelErr so you can inspect it
}
```
Nevertheless, I don't think the model should contain errors. Model should represent the objects of the application. Error signals the incorrect state. So ideally, the package containing the logic should have errors specific to its logic.

Also, the issue I have with this kind of design is that your service knows about the transport layer, which kind of defeats the whole purpose of splitting it into handler/service/db. Handler is responsible for communicating values. Service is responsible for the logic and the db holds the state. If you know that your app is always going to be using http, what is the purpose of the handler/service split?

I know apps that can be triggered either through web sockets or HTTP requests. Then it would re-use service logic, but the error code returned would mean nothing to the websocket (because it is basically just a TCP stream wrapper)

Reply:
Oh, I got my mistake. Thank you :)
Please check how I changed working with errors for this method. Now layers are less bound between each other. If you approve this, I'll change error handling for all other methods in this app